File: MiniDOMOffice/Magazzino/Inventario/OggettiInventariati.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using static minidom.Store;


namespace minidom
{
    namespace repositories
    {

        /// <summary>
        /// Repository di oggetti <see cref="OggettoInventariato"/>
        /// </summary>
        [Serializable]
        public partial class COggettiInventariatiClass 
            : CModulesClass<OggettoInventariato>
        {

            /// <summary>
            /// Costruttore
            /// </summary>
            public COggettiInventariatiClass() 
                : base("modOfficeOggettiInventariati", typeof(OggettoInventariatoCursor), 0)
            {
            }

            //protected override Sistema.CModule CreateModuleInfo()
            //{
            //    var m = base.CreateModuleInfo();
            //    m.Parent = Module;
            //    m.Visible = true;
            //    m.Save();
            //    return m;
            //}

            /// <summary>
            /// Restituisce una stringa che descrive il formato
            /// </summary>
            /// <param name="value"></param>
            /// <returns></returns>
            public string FormatStatoAttuale(StatoOggettoInventariato value)
            {
                switch (value)
                {
                    case StatoOggettoInventariato.Sconosciuto: return "Sconosciuto";
                    case StatoOggettoInventariato.Funzionante: return "Funzionante";
                    case StatoOggettoInventariato.ConQualcheDifetto: return "Con qualche difetto";
                    case StatoOggettoInventariato.DaRiparare: return "Da Riparare";
                    case StatoOggettoInventariato.ValutazioneRiparazione: return "Valutare Riparazione";
                    case StatoOggettoInventariato.NonRiparabile: return "Non riparabile";
                    case StatoOggettoInventariato.Dismesso: return "Dismesso";
                    default: return DMD.Strings.Combine("[", value, "]");
                }
            }

            /// <summary>
            /// Restituisce la collezione degli oggetti che rispettano il pattern
            /// </summary>
            /// <param name="text"></param>
            /// <returns></returns>
            public CCollection<OggettoInventariato> Find(string text)
            {
                var ret = new CCollection<OggettoInventariato>();
                text = Strings.Trim(text);

                if (string.IsNullOrEmpty(text))
                    return ret;

                using (var cursor = new OggettoInventariatoCursor())
                {
                    // cerciahimo prima in base al codice
                    cursor.WhereClauses *= (
                                             cursor.Field ("Seriale").StartsWith(text) 
                                           + cursor.Field("CodiceRFID").StartWidth(text) 
                                           + cursor.Field("Codice").StartWidth(text)
                                           );
                    while (cursor.Read())
                    {
                        ret.Add(cursor.Item);
                    }

                    cursor.Reset1();
                    cursor.WhereClauses = DBCursorField.True;
                    if (ret.Count == 0)
                    {
                        // Se non abbiamo risultati cerchiamo in base al nome
                        cursor.Clear();
                        cursor.Nome.Value = text;
                        cursor.Nome.Operator = OP.OP_LIKE;
                        while (cursor.Read())
                        {
                            ret.Add(cursor.Item);
                        }
                    }

                    cursor.Reset1();
                    if (cursor.Count() == 0)
                    {
                        // Se non abbiamo risultati cerchiamo il base al nome articolo
                        cursor.Clear();
                        cursor.NomeArticolo.Value = text;
                        cursor.NomeArticolo.Operator = OP.OP_LIKE;
                        while (cursor.Read())
                        {
                            ret.Add(cursor.Item);
                        }
                    }

                }

                return ret;
            }

            /// <summary>
            /// Restituisce l'oggetto in base al nome
            /// </summary>
            /// <param name="name"></param>
            /// <returns></returns>
            public OggettoInventariato GetItemByName(string name)
            {
                OggettoInventariato ret = null;

                name = DMD.Strings.Trim(name);
                
                if (string.IsNullOrEmpty(name))
                    return null;

                using (var cursor = new OggettoInventariatoCursor())
                {
                    cursor.Stato.Value = ObjectStatus.OBJECT_VALID;
                    cursor.IgnoreRights = true;
                    cursor.Nome.Value = name;
                    cursor.StatoAttuale.Value = StatoOggettoInventariato.Funzionante;
                    ret = cursor.Item;
                    cursor.Reset1();
                    if (ret is null)
                    {
                        cursor.StatoAttuale.Clear();
                        cursor.StatoAttuale.SortOrder = SortEnum.SORT_ASC;
                        ret = cursor.Item;
                    }

                }

                return ret;
            }
        }
    }

    public partial class Store
    {
        private static COggettiInventariatiClass m_OggettiInventariati = null;

        /// <summary>
        /// Repository di oggetti <see cref="OggettoInventariato"/>
        /// </summary>
        public static COggettiInventariatiClass OggettiInventariati
        {
            get
            {
                if (m_OggettiInventariati is null)
                    m_OggettiInventariati = new COggettiInventariatiClass();
                return m_OggettiInventariati;
            }
        }
    }
}

File: MiniDOMOffice/Magazzino/Inventario/OggettiInventariati.vb
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Anagrafica
Imports minidom.Office
Imports minidom.Internals.Office

Namespace Internals.Office

    Public Class COggettiInventariatiClass
        Inherits CModulesClass(Of OggettoInventariato)

        Public Sub New()
            MyBase.New("modOfficeOggettiInventariati", GetType(OggettoInventariatoCursor), 0)
        End Sub

        Protected Overrides Function CreateModuleInfo() As CModule
            Dim m As CModule = MyBase.CreateModuleInfo()
            m.Parent = minidom.Office.Module
            m.Visible = True
            m.Save()
            Return m
        End Function

        Public Function FormatStatoAttuale(ByVal value As StatoOggettoInventariato) As String
            Select value
                Case StatoOggettoInventariato.Sconosciuto : Return "Sconosciuto"
                Case StatoOggettoInventariato.Funzionante : Return "Funzionante"
                Case StatoOggettoInventariato.ConQualcheDifetto : Return "Con qualche difetto"
                Case StatoOggettoInventariato.DaRiparare : Return "Da Riparare"
                Case StatoOggettoInventariato.ValutazioneRiparazione : Return "Valutare Riparazione"
                Case StatoOggettoInventariato.NonRiparabile : Return "Non riparabile"
                Case StatoOggettoInventariato.Dismesso : Return "Dismesso"
                Case Else : Return "[" & value & "]"
            End Select
        End Function

        Public Function Find(ByVal text As String) As CCollection(Of OggettoInventariato)
            Dim ret As New CCollection(Of OggettoInventariato)
            text = Trim(text)
            If (text = "") Then Return ret
            Dim cursor As New OggettoInventariatoCursor
            'cerciahimo prima in base al codice
            cursor.WhereClauses.Add("[Seriale] Like '" & Replace(text, "'", "''") & "%' OR [CodiceRFID] Like '" & Replace(text, "'", "''") & "%' OR [Codice] Like '" & Replace(text, "'", "''") & "%'")
            While (Not cursor.EOF)
                ret.Add(cursor.Item)
                cursor.MoveNext()
            End While
            cursor.Reset1()
            cursor.WhereClauses.Clear()

            If (ret.Count = 0) Then
                'Se non abbiamo risultati cerchiamo in base al nome
                cursor.Clear()
                cursor.Nome.Value = text & "%"
                cursor.Nome.Operator = OP.OP_LIKE
                While (Not cursor.EOF)
                    ret.Add(cursor.Item)
                    cursor.MoveNext()
                End While
            End If
            cursor.Reset1()

            If (cursor.Count = 0) Then
                'Se non abbiamo risultati cerchiamo il base al nome articolo
                cursor.Clear()
                cursor.NomeArticolo.Value = text & "%"
                cursor.NomeArticolo.Operator = OP.OP_LIKE
                While (Not cursor.EOF)
                    ret.Add(cursor.Item)
                    cursor.MoveNext()
                End While
            End If

            If (cursor IsNot Nothing) Then cursor.Dispose() : cursor = Nothing

            Return ret
        End Function

        Public Function GetItemByName(ByVal name As String) As OggettoInventariato
            name = Strings.Trim(name)
            If (name = "") Then Return Nothing
            Dim cursor As New OggettoInventariatoCursor
            Dim ret As OggettoInventariato = Nothing
            cursor.Stato.Value = ObjectStatus.OBJECT_VALID
            cursor.IgnoreRights = True
            cursor.Nome.Value = name
            cursor.StatoAttuale.Value = StatoOggettoInventariato.Funzionante
            ret = cursor.Item
            cursor.Reset1()
            If (ret Is Nothing) Then
                cursor.StatoAttuale.Clear()
                cursor.StatoAttuale.SortOrder = SortEnum.SORT_ASC
                ret = cursor.Item
            End If
            If (cursor IsNot Nothing) Then cursor.Dispose() : cursor = Nothing
            Return ret
        End Function

    End Class


End Namespace

Partial Public Class Office
     
    Private Shared m_OggettiInventariati As COggettiInventariatiClass = Nothing

    Public Shared ReadOnly Property OggettiInventariati As COggettiInventariatiClass
        Get
            If (m_OggettiInventariati Is Nothing) Then m_OggettiInventariati = New COggettiInventariatiClass
            Return m_OggettiInventariati
        End Get
    End Property




   
End Class


